File: JS/MaximumTwinSumofLinkedList.js
var pairSum = function(head) {
    let slow = head;
    let fast = head;

    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
    }

    let prev = null;
    let curr = slow;

    while (curr) {
        let next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
    }

    let ans = 0;
    let first = head;
    let second = prev;

    while (second) {
        ans = Math.max(ans, first.val + second.val);
        first = first.next;
        second = second.next;
    }

    return ans;
};

File: Java/MaximumTwinSumOfLinkedList.java
import java.util.*;

// Step 1: Find the mid node - slow/fast pointers
// Step 2: Reverse the second half of list
// Step 3: Compare correspondingtwin elements

class ListNode {
    public int pairSum(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode secondHalf = reverse(slow);
        int maxSum = 0;
        ListNode firstHalf = head;

        while (secondHalf != null) {
            maxSum = Math.max(maxSum, firstHalf.val + secondHalf.val);
            firstHalf = firstHalf.next;
            secondHalf = secondHalf.next;
        }

        return maxSum;
    }

    private ListNode reverse(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;

        while (curr != null) {
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return prev;
    }
}
public class MaximumTwinSumOfLinkedList {
    public static void main(String[] args) {
        
    }
}

